network/src/transport.rs

```diff
@@ -397,3 +420,92 @@ pub fn build_tcp_transport(
         .with_timeout(TRANSPORT_TIMEOUT)
         .boxed()
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        protocols::wire::handshake::v1::{HandshakeMsg, MessagingProtocolVersion},
+        transport::perform_handshake,
+        ProtocolId,
+    };
+    use futures::{executor::block_on, future::join};
+    use libra_config::network_id::NetworkId;
+    use libra_network_address::NetworkAddress;
+    use libra_types::PeerId;
+    use memsocket::MemorySocket;
+    use netcore::transport::ConnectionOrigin;
+
+    fn build_test_connection() -> (MemorySocket, MemorySocket) {
+        MemorySocket::new_pair()
+    }
+
+    #[test]
+    fn handshake_network_id_mismatch() {
+        let peer_id_1 = PeerId::from_hex_literal("0x00").unwrap();
+        let peer_id_2 = PeerId::from_hex_literal("0x01").unwrap();
+        let network_addr_1 = NetworkAddress::mock();
+        let network_addr_2 = NetworkAddress::mock();
+        let (outbound, inbound) = build_test_connection();
+        let validator_network = NetworkId::Validator;
+        let public_network = NetworkId::Public;
+
+        // Create client and server handshake messages.
+        let mut server_handshake = HandshakeMsg::new(validator_network);
+        server_handshake.add(
+            MessagingProtocolVersion::V1,
+            [
+                ProtocolId::ConsensusDirectSend,
+                ProtocolId::MempoolDirectSend,
+            ]
```

Comment:
we don't need to actually be this verbose, so maybe we can just strip out a lot of boiler plate stuff like this?

Reply:
I stole this all from https://github.com/libra/libra/blob/1df4359a5fc4d2c82b636152987071b57021c768/network/src/protocols/identity.rs#L89-L138

I'll take a round to simplify it.

The only reason I added the test here and not in the identity protocol, is because it doesn't cover the application level check on equality of networkIds.